teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md

```diff
@@ -171,12 +185,277 @@ In addition to grouping related resources, Tekton Results provides long term sto
+	RunsStatus []PipelineRunRunStatus `json:"runsStatus,omitempty"`
+	...
+}
+```
+
+See the [Beta API](#beta-api) and [V1 API](#v1-api) sections below for more information on the
+migration plan.
+
+#### Example
+
+This is an example `PipelineRun` status as provided in the [documentation][example]:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  ...
+spec:
+    ...
+status:
+    completionTime: "2020-05-04T02:19:14Z"
+    conditions:
+      - lastTransitionTime: "2020-05-04T02:19:14Z"
+        message: "Tasks Completed: 4, Skipped: 0"
+        reason: Succeeded
+        status: "True"
+        type: Succeeded
+    startTime: "2020-05-04T02:00:11Z"
+    taskRuns:
+      triggers-release-nightly-build:
+        pipelineTaskName: build
+        status:
+          completionTime: "2020-05-04T02:10:49Z"
+          conditions:
+            - lastTransitionTime: "2020-05-04T02:10:49Z"
+              message: All Steps have completed executing
+              reason: Succeeded
+              status: "True"
+              type: Succeeded
+          podName: triggers-release-nightly-build-pod
+          resourcesResult:
+            - key: commit
+              resourceRef:
+                name: git-source-triggers
+              value: 9ab5a1234166a89db352afa28f499d596ebb48db
+          startTime: "2020-05-04T02:05:07Z"
+          steps:
+            - container: step-build
+              imageID: docker-pullable://golang@sha256:a90f2671330831830e229c3554ce118009681ef88af659cd98bfafd13d5594f9
+              name: build
+              terminated:
+                containerID: docker://6b6471f501f59dbb7849f5cdde200f4eeb64302b862a27af68821a7fb2c25860
+                exitCode: 0
+                finishedAt: "2020-05-04T02:10:45Z"
+                reason: Completed
+                startedAt: "2020-05-04T02:06:24Z"
+```
+
+Taking the above example, this will be the new minimal `PipelineRun` status:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  ...
+spec:
+    ...
+status:
+    completionTime: "2020-05-04T02:19:14Z"
+    conditions:
+      - lastTransitionTime: "2020-05-04T02:19:14Z"
+        message: "Tasks Completed: 4, Skipped: 0"
+        reason: Succeeded
+        status: "True"
+        type: Succeeded
+    startTime: "2020-05-04T02:00:11Z"
+    taskRunsStatus:
+      - taskRunName: triggers-release-nightly-build
+        conditions:
+          - lastTransitionTime: "2020-05-04T02:10:49Z"
+            message: All Steps have completed executing
+            reason: Succeeded
+            status: "True"
+            type: Succeeded
+```
+
+#### Beta API
+
+Because the `PipelineRun` status is part of the [Tekton Pipelines API][api-definition],
+this API change is not backwards compatible. To support migration, we will add a new 
+behavior flag - `disable-embedded-status` - used to configure whether `PipelineRuns`
+should contain the embedded status or minimal status of its `TaskRuns` and `Runs`.
```

Comment:
(assuming this means xor...) It might be a softer transition if we populate both fields in API responses for a phased rollout, which might mean we will want this flag to be a string enum rather than a bool.

 i.e.

1. (what we have today) only populate full embedded status
2. populate full embedded status + minimal embedded status - this would give clients the chance to migrating to the new field w/o breaking existing behavior.
3. only populate minimal embedded status

Reply:
2 birds with one stone for the flag name bikeshed:

```
--embedded-status=[ full | both | minimal ]
```

Reply:
I don't think we should support both, because it cause more duplication and worsen the problems we're trying to solve. However, I like the approach of using string enums instead of bools. And by using strings instead of bools, we can expand the field to support `both` later if we receive feedback from users that they need it. Updated the flag to be:

```
--embedded-status=[ full | minimal ]
```

Reply:
Supporting both is unfortunately going to be a cost of migration with no down time :( 

Otherwise you're going to have a period of time where clients expecting the full status are going to suddenly break because the server is now returning minimal, or clients expecting minimal status don't work yet because the server is still returning full. There needs to be a period of time where clients can use either so they can migrate.

Note there's actually not a strict requirement to **store** both in etcd - we just need to make sure that API responses populate both, but idk how feasible this would be to implement in only the API without moving to an [API aggregation model ](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#api-server-aggregation)

Reply:
sounds great, updated the TEP to provide `both` as an option too, thanks!

https://github.com/jerop/community/blob/tep-0100-proposal/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md#beta-api